internal/githubapi/roundtripper.go

```diff
@@ -0,0 +1,129 @@
+package githubapi
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/v44/github"
+	"github.com/ossf/criticality_score/internal/retry"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	githubErrorIdSearch = "\"error_500\""
+)
+
+var (
+	issueCommentsRe = regexp.MustCompile("^repos/[^/]+/[^/]+/issues/comments$")
+)
+
+func NewRoundTripper(rt http.RoundTripper, logger *log.Logger) http.RoundTripper {
+	s := &strategies{logger: logger}
+	return retry.NewRoundTripper(rt,
+		retry.InitialDelay(2*time.Minute),
+		retry.RetryAfter(s.RetryAfter),
+		retry.Strategy(s.SecondaryRateLimit),
+		retry.Strategy(s.ServerError400),
+		retry.Strategy(s.ServerError),
+	)
+}
+
+type strategies struct {
+	logger *log.Logger
+}
+
+func respBodyContains(r *http.Response, search string) (bool, error) {
+	data, err := ioutil.ReadAll(r.Body)
+	defer r.Body.Close()
+	r.Body = ioutil.NopCloser(bytes.NewBuffer(data))
+	if err != nil {
+		return false, err
+	}
+	return bytes.Contains(data, []byte(search)), nil
+}
+
+// ServerError implements retry.RetryStrategyFn
+func (s *strategies) ServerError(r *http.Response) (retry.RetryStrategy, error) {
+	if r.StatusCode < 500 || 600 <= r.StatusCode {
+		return retry.NoRetry, nil
+	}
+	s.logger.WithField("status", r.Status).Warn("5xx: detected")
+	if issueCommentsRe.MatchString(strings.Trim(r.Request.URL.Path, "/")) {
+		s.logger.Warn("Ignoring /repos/X/Y/issues/comments url.")
+		// If the req url was /repos/[owner]/[name]/issues/comments pass the
+		// error through as it is likely a GitHub bug.
+		return retry.NoRetry, nil
+	}
+	return retry.RetryImmediate, nil
+
+}
+
+// ServerError400 implements retry.RetryStrategyFn
+func (s *strategies) ServerError400(r *http.Response) (retry.RetryStrategy, error) {
+	if r.StatusCode != http.StatusBadRequest {
+		return retry.NoRetry, nil
+	}
+	s.logger.Warn("400: bad request detected")
+	if r.Header.Get("Content-Type") != "text/html" {
+		return retry.NoRetry, nil
+	}
+	s.logger.Debug("It's a text/html doc")
+	if isError, err := respBodyContains(r, githubErrorIdSearch); isError {
+		s.logger.Debug("Found target string - assuming 500.")
+		return retry.RetryImmediate, nil
+	} else {
+		return retry.NoRetry, err
+	}
+}
+
+// SecondaryRateLimit implements retry.RetryStrategyFn
+func (s *strategies) SecondaryRateLimit(r *http.Response) (retry.RetryStrategy, error) {
+	if r.StatusCode != http.StatusForbidden {
+		return retry.NoRetry, nil
+	}
+	s.logger.Warn("403: forbidden detected")
+	errorResponse := &github.ErrorResponse{Response: r}
+	data, err := ioutil.ReadAll(r.Body)
+	r.Body.Close()
+	r.Body = ioutil.NopCloser(bytes.NewBuffer(data))
+	if err != nil || data == nil {
+		s.logger.WithFields(
+			log.Fields{
+				"error":    err,
+				"data_nil": (data == nil),
+			}).Warn("ReadAll failed.")
+		return retry.NoRetry, err
+	}
+	json.Unmarshal(data, errorResponse)
```

Comment:
Should we be checking for error here?

Reply:
This code is copied from the go-github repo. If it's not valid JS then this response should fall through to the other error handling.

Reply:
Added a comment to make this clearer.

Reply:
ack, thanks!